File: backend-ex/controllers/cartController.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserCart = exports.updateCart = exports.addToCart = void 0;
const cartModel_1 = require("../models/cartModel");
const addToCart = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const { userId, itemId, size } = req.body;
        const existingItem = yield (0, cartModel_1.findCartItem)(userId, itemId, size);
        if (existingItem) {
            yield (0, cartModel_1.incrementCartItemQuantity)(userId, itemId, size);
            return res.json({ message: 'Quantity updated.' });
        }
        else {
            yield (0, cartModel_1.insertCartItem)(userId, itemId, size);
            return res.json({ message: 'Item added to cart.' });
        }
    }
    catch (error) {
        console.error('Error in addToCart:', error);
        res.status(500).json({ error: 'Server error' });
    }
});
exports.addToCart = addToCart;
const updateCart = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const { userId, itemId, size, quantity } = req.body;
        const existingItem = yield (0, cartModel_1.findCartItem)(userId, itemId, size);
        if (!existingItem) {
            return res.status(404).json({ error: 'Item not found in cart.' });
        }
        yield (0, cartModel_1.updateCartItemQuantity)(userId, itemId, size, quantity);
        return res.json({ message: 'Quantity updated successfully.' });
    }
    catch (error) {
        console.error('Error in updateCart:', error);
        res.status(500).json({ error: 'Server error' });
    }
});
exports.updateCart = updateCart;
const getUserCart = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const userId = Number(req.params.userId);
        if (isNaN(userId)) {
            return res.status(400).json({ error: 'Invalid userId parameter.' });
        }
        const cartItems = yield (0, cartModel_1.getCartItemsByUser)(userId);
        return res.json({ cart: cartItems });
    }
    catch (error) {
        console.error('Error in getUserCart:', error);
        res.status(500).json({ error: 'Server error' });
    }
});
exports.getUserCart = getUserCart;

File: backend-go/controllers/cart_controller.go
package controllers

import (
	"fmt"
	"net/http"

	"github.com/avalent1/snapshop/repositories"
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type CartRequest struct {
	ItemID   int    `json:"itemId"`
	Size     string `json:"size"`
	Quantity int    `json:"quantity"`
}

// POST /cart/add
func AddToCart(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req CartRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid input"})
			return
		}

		userID := c.GetInt("userId")

		existingItem, err := repositories.FindCartItem(userID, req.ItemID, req.Size)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
			return
		}

		if existingItem != nil {
			if err := repositories.IncrementCartItemQuantity(userID, req.ItemID, req.Size); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to increment quantity"})
				return
			}
			c.JSON(http.StatusOK, gin.H{"message": "Quantity updated."})
		} else {
			if err := repositories.InsertCartItem(userID, req.ItemID, req.Size); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add item to cart"})
				return
			}
			c.JSON(http.StatusOK, gin.H{"message": "Item added to cart."})
		}
	}
}

// PUT /cart/update
func UpdateCart(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req CartRequest
		if err := c.ShouldBindJSON(&req); err != nil || req.Quantity <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid input"})
			return
		}

		userID := c.GetInt("userId")

		existingItem, err := repositories.FindCartItem(userID, req.ItemID, req.Size)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
			return
		}
		if existingItem == nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "Item not found in cart."})
			return
		}

		if err := repositories.UpdateCartItemQuantity(userID, req.ItemID, req.Size, req.Quantity); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update quantity"})
			return
		}

		c.JSON(http.StatusOK, gin.H{"message": "Quantity updated successfully."})
	}
}

// POST /cart/get
func GetUserCart(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := c.GetInt("userId")
		fmt.Println("user id je: ", userID)
		detailedItems, err := repositories.GetDetailedCartItems(userID)
		//cartItems, err := repositories.GetCartItemsByUser(db, userID)
		if err != nil {
			fmt.Printf("Error retrieving cart items: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Server error"})
			return
		}

		c.JSON(http.StatusOK, gin.H{"cartData": detailedItems})
	}
}

// DELETE /cart/remove
func RemoveFromCart(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req CartRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid input"})
			return
		}

		userID := c.GetInt("userId")

		existingItem, err := repositories.FindCartItem(userID, req.ItemID, req.Size)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
			return
		}
		if existingItem == nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "Item not found in cart."})
			return
		}

		if err := repositories.DeleteCartItem(userID, req.ItemID, req.Size); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete item"})
			return
		}

		c.JSON(http.StatusOK, gin.H{"message": "Item removed from cart."})
	}
}
